File: lib/features/business/domain/entities/site_config.dart
import 'package:equatable/equatable.dart';

import 'site_cta_target.dart';

/// Public storefront theme (CSS variables on shop).
class SiteTheme extends Equatable {
  const SiteTheme({
    this.primary = '#1c1917',
    this.accent = '#ff6b35',
    this.background = '#ffffff',
    this.text = '#111827',
  });

  final String primary;
  final String accent;
  final String background;
  final String text;

  SiteTheme copyWith({
    String? primary,
    String? accent,
    String? background,
    String? text,
  }) =>
      SiteTheme(
        primary: primary ?? this.primary,
        accent: accent ?? this.accent,
        background: background ?? this.background,
        text: text ?? this.text,
      );

  @override
  List<Object?> get props => [primary, accent, background, text];
}

class SiteGalleryItem extends Equatable {
  const SiteGalleryItem({
    this.imageUrl,
    this.youtubeUrl,
    this.caption,
  });

  final String? imageUrl;
  final String? youtubeUrl;
  final String? caption;

  SiteGalleryItem copyWith({
    String? imageUrl,
    String? youtubeUrl,
    String? caption,
  }) =>
      SiteGalleryItem(
        imageUrl: imageUrl ?? this.imageUrl,
        youtubeUrl: youtubeUrl ?? this.youtubeUrl,
        caption: caption ?? this.caption,
      );

  @override
  List<Object?> get props => [imageUrl, youtubeUrl, caption];
}

/// One configurable block on the public shop (hero, products, gallery, …).
class SiteSectionConfig extends Equatable {
  const SiteSectionConfig({
    required this.id,
    this.enabled = true,
    this.title,
    this.description,
    this.navLabel,
    this.imageUrl,
    this.useProfileCover = true,
    this.galleryItems = const [],
    this.ctaLabel,
    this.secondaryCtaLabel,
    this.ctaTarget,
    this.secondaryCtaTarget,
    this.trustBullets = const [],
  });

  final String id;
  final bool enabled;
  final String? title;
  final String? description;
  final String? navLabel;
  final String? imageUrl;
  final bool useProfileCover;
  final List<SiteGalleryItem> galleryItems;
  /// Primary button label (hero, contact WhatsApp, etc.).
  final String? ctaLabel;
  /// Hero secondary button (e.g. services).
  final String? secondaryCtaLabel;
  /// Where the primary CTA navigates (products, contact, whatsapp, …).
  final SiteCtaTarget? ctaTarget;
  final SiteCtaTarget? secondaryCtaTarget;
  /// Hero trust bullets (one per line in admin).
  final List<String> trustBullets;

  SiteSectionConfig copyWith({
    bool? enabled,
    String? title,
    String? description,
    String? navLabel,
    String? imageUrl,
    bool? useProfileCover,
    List<SiteGalleryItem>? galleryItems,
    String? ctaLabel,
    String? secondaryCtaLabel,
    SiteCtaTarget? ctaTarget,
    SiteCtaTarget? secondaryCtaTarget,
    List<String>? trustBullets,
  }) =>
      SiteSectionConfig(
        id: id,
        enabled: enabled ?? this.enabled,
        title: title ?? this.title,
        description: description ?? this.description,
        navLabel: navLabel ?? this.navLabel,
        imageUrl: imageUrl ?? this.imageUrl,
        useProfileCover: useProfileCover ?? this.useProfileCover,
        galleryItems: galleryItems ?? this.galleryItems,
        ctaLabel: ctaLabel ?? this.ctaLabel,
        secondaryCtaLabel: secondaryCtaLabel ?? this.secondaryCtaLabel,
        ctaTarget: ctaTarget ?? this.ctaTarget,
        secondaryCtaTarget: secondaryCtaTarget ?? this.secondaryCtaTarget,
        trustBullets: trustBullets ?? this.trustBullets,
      );

  @override
  List<Object?> get props => [
        id,
        enabled,
        title,
        description,
        navLabel,
        imageUrl,
        useProfileCover,
        galleryItems,
        ctaLabel,
        secondaryCtaLabel,
        ctaTarget,
        secondaryCtaTarget,
        trustBullets,
      ];
}

class SiteSocialLink extends Equatable {
  const SiteSocialLink({required this.platform, required this.url});

  final String platform;
  final String url;

  @override
  List<Object?> get props => [platform, url];
}

class SiteConfig extends Equatable {
  const SiteConfig({
    this.theme = const SiteTheme(),
    this.layout = 'default',
    this.sections = const [],
    this.socials = const [],
    this.footerShowLocation = true,
    this.footerShowPhone = true,
    this.footerShowWhatsApp = true,
  });

  final SiteTheme theme;
  final String layout;
  final List<SiteSectionConfig> sections;
  final List<SiteSocialLink> socials;
  final bool footerShowLocation;
  final bool footerShowPhone;
  final bool footerShowWhatsApp;

  static const sectionHero = 'hero';
  static const sectionOffers = 'offers';
  static const sectionContests = 'contests';
  static const sectionJobOpenings = 'jobOpenings';
  static const sectionProducts = 'products';
  static const sectionServices = 'services';
  static const sectionAbout = 'about';
  static const sectionGallery = 'gallery';
  static const sectionContact = 'contact';

  /// Max images/videos per gallery section (admin + public API).
  static const maxGalleryItems = 8;

  static const socialPlatforms = [
    'facebook',
    'instagram',
    'tiktok',
    'youtube',
    'linkedin',
    'x',
    'whatsapp',
  ];

  static SiteConfig defaults() => SiteConfig(
        sections: [
          const SiteSectionConfig(
            id: sectionHero,
            enabled: true,
            useProfileCover: true,
          ),
          const SiteSectionConfig(id: sectionOffers, enabled: true, title: 'Ofertat'),
          const SiteSectionConfig(id: sectionContests, enabled: true, title: 'Konkurset'),
          const SiteSectionConfig(
            id: sectionJobOpenings,
            enabled: true,
            title: 'Konkurse pune',
          ),
          const SiteSectionConfig(
            id: sectionProducts,
            enabled: true,
            title: 'Produktet',
          ),
          const SiteSectionConfig(
            id: sectionServices,
            enabled: true,
            title: 'Shërbimet',
          ),
          const SiteSectionConfig(id: sectionAbout, enabled: true, title: 'Rreth nesh'),
          const SiteSectionConfig(id: sectionGallery, enabled: true, title: 'Galeria'),
          const SiteSectionConfig(
            id: sectionContact,
            enabled: true,
            title: 'Kontakt',
          ),
        ],
      );

  SiteSectionConfig? section(String id) {
    for (final s in sections) {
      if (s.id == id) return s;
    }
    return null;
  }

  SiteConfig copyWith({
    SiteTheme? theme,
    String? layout,
    List<SiteSectionConfig>? sections,
    List<SiteSocialLink>? socials,
    bool? footerShowLocation,
    bool? footerShowPhone,
    bool? footerShowWhatsApp,
  }) =>
      SiteConfig(
        theme: theme ?? this.theme,
        layout: layout ?? this.layout,
        sections: sections ?? this.sections,
        socials: socials ?? this.socials,
        footerShowLocation: footerShowLocation ?? this.footerShowLocation,
        footerShowPhone: footerShowPhone ?? this.footerShowPhone,
        footerShowWhatsApp: footerShowWhatsApp ?? this.footerShowWhatsApp,
      );

  SiteConfig mergeWithDefaults() {
    final defaults = SiteConfig.defaults();
    final byId = {for (final s in sections) s.id: s};
    final merged = defaults.sections.map((d) => byId[d.id] ?? d).toList();
    for (final s in sections) {
      if (!merged.any((m) => m.id == s.id)) merged.add(s);
    }
    return copyWith(sections: merged);
  }

  @override
  List<Object?> get props =>
      [theme, layout, sections, socials, footerShowLocation, footerShowPhone, footerShowWhatsApp];
}

File: functions/siteConfig.js
const ALLOWED_SECTION_IDS = new Set([
  'hero',
  'offers',
  'contests',
  'jobOpenings',
  'products',
  'services',
  'about',
  'gallery',
  'contact',
]);

const ALLOWED_CTA_TARGETS = new Set([
  'products',
  'services',
  'contact',
  'offers',
  'contests',
  'jobOpenings',
  'whatsapp',
]);

const ctaPresets = require('./siteCtaPresets');

const ALLOWED_SOCIAL = new Set([
  'facebook',
  'instagram',
  'tiktok',
  'youtube',
  'linkedin',
  'x',
  'whatsapp',
]);

const HEX_COLOR = /^#([0-9a-fA-F]{3}|[0-9a-fA-F]{6}|[0-9a-fA-F]{8})$/;

const MAX_GALLERY_ITEMS = 8;

function clip(str, max) {
  if (str == null) return '';
  return String(str).trim().slice(0, max);
}

function sanitizeColor(value, fallback) {
  const v = clip(value, 9);
  return HEX_COLOR.test(v) ? v : fallback;
}

function sanitizeText(value, max = 500) {
  return clip(value, max).replace(/<[^>]*>/g, '');
}

function sanitizeUrl(value, max = 500) {
  const u = clip(value, max);
  if (!u) return '';
  if (/^https?:\/\//i.test(u)) return u;
  if (/^www\./i.test(u)) return `https://${u}`;
  return u;
}

function sanitizeYoutube(value) {
  const u = clip(value, 200);
  if (!u) return '';
  if (/^https?:\/\//i.test(u) || u.includes('youtube.com') || u.includes('youtu.be')) {
    return u;
  }
  const id = u.replace(/[^a-zA-Z0-9_-]/g, '').slice(0, 20);
  if (!id) return '';
  return `https://www.youtube.com/embed/${id}`;
}

function defaultSiteConfig() {
  return {
    theme: {
      primary: '#1c1917',
      accent: '#ff6b35',
      background: '#ffffff',
      text: '#1c1917',
    },
    layout: 'default',
    sections: [
      { id: 'hero', enabled: true, useProfileCover: true },
      { id: 'offers', enabled: true, title: 'Ofertat' },
      { id: 'contests', enabled: true, title: 'Konkurset' },
      { id: 'jobOpenings', enabled: true, title: 'Konkurse pune' },
      { id: 'products', enabled: true, title: 'Produktet' },
      { id: 'services', enabled: true, title: 'Shërbimet' },
      { id: 'about', enabled: true, title: 'Rreth nesh' },
      { id: 'gallery', enabled: true, title: 'Galeria', galleryItems: [] },
      { id: 'contact', enabled: true, title: 'Kontakt' },
    ],
    socials: [],
    footerShowLocation: true,
    footerShowPhone: true,
    footerShowWhatsApp: true,
  };
}

function mergeSections(saved) {
  const defaults = defaultSiteConfig().sections;
  const byId = new Map();
  if (Array.isArray(saved)) {
    for (const s of saved) {
      if (s && s.id && ALLOWED_SECTION_IDS.has(s.id)) byId.set(s.id, s);
    }
  }
  return defaults.map((d) => ({ ...d, ...(byId.get(d.id) || {}) }));
}

/**
 * Sanitize siteConfig from Firestore for public API.
 */
function sanitizeCtaTarget(value) {
  const v = clip(value, 20);
  return ALLOWED_CTA_TARGETS.has(v) ? v : '';
}

function publicSiteConfig(raw, businessType = '') {
  if (!raw || typeof raw !== 'object') {
    const base = defaultSiteConfig();
    return {
      ...base,
      sections: ctaPresets.enrichSections(base.sections, businessType),
    };
  }

  const themeIn = raw.theme && typeof raw.theme === 'object' ? raw.theme : {};
  const mergedSections = mergeSections(raw.sections);

  const sections = mergedSections.map((s) => {
    const out = {
      id: s.id,
      enabled: s.enabled !== false,
      useProfileCover: s.useProfileCover !== false,
    };
    const title = sanitizeText(s.title, 120);
    const description = sanitizeText(s.description, 800);
    const navLabel = sanitizeText(s.navLabel, 40);
    const ctaLabel = sanitizeText(s.ctaLabel, 60);
    const secondaryCtaLabel = sanitizeText(s.secondaryCtaLabel, 60);
    if (title) out.title = title;
    if (description) out.description = description;
    if (navLabel) out.navLabel = navLabel;
    if (ctaLabel) out.ctaLabel = ctaLabel;
    if (secondaryCtaLabel) out.secondaryCtaLabel = secondaryCtaLabel;
    const ctaTarget = sanitizeCtaTarget(s.ctaTarget);
    const secondaryCtaTarget = sanitizeCtaTarget(s.secondaryCtaTarget);
    if (ctaTarget) out.ctaTarget = ctaTarget;
    if (secondaryCtaTarget) out.secondaryCtaTarget = secondaryCtaTarget;
    if (s.id === 'hero' && Array.isArray(s.trustBullets)) {
      out.trustBullets = s.trustBullets
        .map((t) => sanitizeText(t, 80))
        .filter(Boolean)
        .slice(0, 5);
    }
    if (s.id === 'hero') {
      const imageUrl = sanitizeUrl(s.imageUrl);
      if (imageUrl) out.imageUrl = imageUrl;
    }
    if (s.id === 'gallery' && Array.isArray(s.galleryItems)) {
      out.galleryItems = s.galleryItems.slice(0, MAX_GALLERY_ITEMS).map((g) => {
        const item = {};
        const img = sanitizeUrl(g?.imageUrl);
        const yt = sanitizeYoutube(g?.youtubeUrl);
        const cap = sanitizeText(g?.caption, 120);
        if (img) item.imageUrl = img;
        if (yt) item.youtubeUrl = yt;
        if (cap) item.caption = cap;
        return item;
      }).filter((g) => g.imageUrl || g.youtubeUrl);
    }
    return out;
  });

  const socials = [];
  if (Array.isArray(raw.socials)) {
    for (const s of raw.socials) {
      const platform = clip(s?.platform, 20).toLowerCase();
      const url = sanitizeUrl(s?.url);
      if (!ALLOWED_SOCIAL.has(platform) || !url) continue;
      socials.push({ platform, url });
    }
  }

  const enrichedSections = ctaPresets.enrichSections(sections, businessType);

  return {
    theme: {
      primary: sanitizeColor(themeIn.primary, '#0a1628'),
      accent: sanitizeColor(themeIn.accent, '#f5c518'),
      background: sanitizeColor(themeIn.background, '#ffffff'),
      text: sanitizeColor(themeIn.text, '#111827'),
    },
    layout: raw.layout === 'wide' ? 'wide' : 'default',
    sections: enrichedSections,
    socials,
    footerShowLocation: raw.footerShowLocation !== false,
    footerShowPhone: raw.footerShowPhone !== false,
    footerShowWhatsApp: raw.footerShowWhatsApp !== false,
  };
}

module.exports = {
  publicSiteConfig,
  defaultSiteConfig,
};
